### DISCUSSION TITLE
How to run must_not query in elastalert2

### DISCUSSION THREAD

**Discussion by ziaul-ict2020** [jertel/elastalert2#1115] [Q&A] [answered]
Hi team,
 I have set the following frequency rule.
```
    name: dtacapp-dev-400-status-alert
    is_enabled: False
    type: frequency
    index: api.log*
    num_events: 20
    timeframe:
      hours: 1
    filter:
    - term: 
        response.statusCode: "400"
    - bool:
        must_not:
        - match:
            routePattern: "/test/v1/test-otp"
    alert:
    - "googlechat"
    googlechat_webhook_url: "https://chat.googleapis.com/v1/..."
    googlechat_format: "card"
    googlechat_header_title: "Warning For Event"
    alert_text_type: alert_text_jinja
    alert_text: |
      Rule-name: {{name}}
      Service: {{service}}
      URL: {{url}}
      httpTotaltime: {{httpTotalTime}}
      msisdn: {{msisdn}}
      Backend_URL: {{http[0].url}}
      http.response code: {{http[0].response_code}}
      Number_Hits: {{num_hits}}
      Number_Matches: {{num_matches}}
```
when I have run the test rule , i have got the below error:
```
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
            To send them but remain verbose, use --verbose instead.
Didn't get any results.
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
INFO:elastalert:1 rules loaded
Traceback (most recent call last):
  File "/usr/local/bin/elastalert-test-rule", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/elastalert/test_rule.py", line 500, in main
    test_instance.run_rule_test()
  File "/usr/local/lib/python3.11/site-packages/elastalert/test_rule.py", line 490, in run_rule_test
    self.run_elastalert(rule_yaml, conf)
  File "/usr/local/lib/python3.11/site-packages/elastalert/test_rule.py", line 409, in run_elastalert
    client.run_rule(rule, self.endtime, self.starttime)
  File "/usr/local/lib/python3.11/site-packages/elastalert/elastalert.py", line 822, in run_rule
    for x in range(len(rule['agg_matches'])):
                       ~~~~^^^^^^^^^^^^^^^
KeyError: 'agg_matches'
```
Actually i want to match only response.statusCode =400 without this document routePattern: "/test/v1/test-otp"

Please help me with your valuable suggestion that how I will be able to resolve this issue.
Thanks
Ziaul

**Comment by jertel** [comment, marked as answer]
Are you aware that the rule you pasted has `is_enabled: False`? This disables the rule.

**Reply by ziaul-ict2020** [reply]
Thank you for catching my mistake. Could you please help me with this rule which I have configured is it correct or not
---
    name: dtacapp-dev-400-status-alert
    is_enabled: True
    type: frequency
    index: api.log*
    num_events: 2
    timeframe:
      minutes: 15
    filter:
    - term: 
        response.statusCode: "400"
    - bool:
        must_not:
        - match:
            routePattern: "/test/v1/test-otp"
    alert:
    - "googlechat"
    googlechat_webhook_url: "https://chat.googleapis.com/v1/spaces/..."
    googlechat_format: "card"
    googlechat_header_title: "Warning For Event"
    alert_text_type: alert_text_jinja
    alert_text: |
      '{{rule_name}}' from '{{start_timestamp}}' to '{{end_timestamp}}': {{num_hits}} / {{num_matches}} hits
      Rule-name: {{name}}
      Service: {{service}}
      URL: {{url}}
      httpTotaltime: {{httpTotalTime}}
      msisdn: {{msisdn}}
      Backend_URL: {{http[0].url}}
      http.response code: {{http[0].response_code}}
      Number_Hits: {{num_hits}}
      Number_Matches: {{num_matches}}
      ---

**Reply by jertel** [reply]
Without running it I can't tell you if it's correct or not. The general syntax looks ok but I could be missing something. Is it failing?

**Reply by ziaul-ict2020** [reply]
It is not showing any error. But whenever i run the test rule every time it shows 0/0 hits

```
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
INFO:elastalert:1 rules loaded
INFO:apscheduler.scheduler:Adding job tentatively -- it will be properly scheduled when the scheduler starts
INFO:elastalert:Queried rule dtacapp-dev-400-status-alert from 2023-02-26 07:21 UTC to 2023-02-26 07:51 UTC: 0 / 0 hits
WARNING:elasticsearch:DELETE https://es.ap-southeast-1.es.amazonaws.com:443/_search/scroll [status:404 request:0.005s]
INFO:elastalert:Queried rule dtacapp-dev-400-status-alert from 2023-02-26 07:51 UTC to 2023-02-26 08:21 UTC: 0 / 0 hits
WARNING:elasticsearch:DELETE https://es.ap-southeast-1.es.amazonaws.com:443/_search/scroll [status:404 request:0.005s]
INFO:elastalert:Queried rule dtacapp-dev-400-status-alert from 2023-02-26 08:21 UTC to 2023-02-26 08:22 UTC: 0 / 0 hits
WARNING:elasticsearch:DELETE https://es.ap-southeast-1.es.amazonaws.com:443/_search/scroll [status:404 request:0.005s]

Would have written the following documents to writeback index (default is elastalert_status):

elastalert_status - {'rule_name': 'dtacapp-dev-400-status-alert', 'endtime': datetime.datetime(2023, 2, 26, 8, 22, 22, 93560, tzinfo=tzutc()), 'starttime': datetime.datetime(2023, 2, 26, 7, 21, 46, 93560, tzinfo=tzutc()), 'matches': 0, 'hits': 0, '@timestamp': datetime.datetime(2023, 2, 26, 8, 22, 22, 644465, tzinfo=tzutc()), 'time_taken': 0.24567699432373047}
```

My actual elasticsearch query is 

```
GET api.log*/_search
{
  "query": {
    "bool": {
      "must": [
        {"term": {
          "response.statusCode": {
            "value": "400"
          }
        }}
      ],
      "must_not": [
        {"match_phrase": {
          "routePattern": "/test/v1/test-otp"
        }}
      ]
    }
  }
}
```

Could you please help me with how I will be able to achieve this on elastalert2?
I will be grateful to you,
Thanks
Ziaul

**Reply by jertel** [reply]
Are you saying that your actual elasticsearch query above does work correctly, and return results? But the ElastAlert 2 rule does not?  If so, you will need to use the debug flag to re-run the rule test so we can see the query being sent to Elasticsearch, and compare the one that works to the one that doesn't.

**Reply by ziaul-ict2020** [reply]
I don't know how to use debug flag. Could you please tell me?

**Reply by jertel** [reply]
In the `logging:` block of the config file, change all the log levels to `DEBUG`:

```
logging:
  version: 1
  incremental: false
  disable_existing_loggers: false
  formatters:
    logline:
      format: '%(asctime)s %(levelname)+8s %(name)+20s %(message)s'

  handlers:
    console:
      class: logging.StreamHandler
      formatter: logline
      level: DEBUG
      stream: ext://sys.stderr

    file:
      class : logging.FileHandler
      formatter: logline
      level: DEBUG
      filename: elastalert.log

  loggers:
    elastalert:
      level: DEBUG
      handlers: []
      propagate: true

    elasticsearch:
      level: DEBUG
      handlers: []
      propagate: true

    elasticsearch.trace:
      level: DEBUG
      handlers: []
      propagate: true

    '':  # root logger
      level: DEBUG
      handlers:
        - console
        - file
      propagate: false
```

**Reply by ziaul-ict2020** [reply]
After more time trying i am success . I have set the below filter.
    filter:
    - bool:
        must:
        - term: 
            "response.statusCode": 
             value: "400" 
        must_not:
        - match_phrase:
            routePattern: "/test/v1/test-otp"
Thank you so much for your nice support.
Now i am facing another thing. I want to send a alert with query time period using aler_text

"At least 5 events occurred between 2023-02-27 04:53 UTC and 2023-02-27 05:53 UTC"

i have tried several time but every time i am failing . It is showing blank when the alert is trigger

my aler configuration is below.

    alert:
    - "googlechat"
    googlechat_webhook_url: "https://chat.googleapis.com/v1/
    googlechat_format: "card"
    googlechat_header_title: "Status_400"
    alert_text_type: alert_text_jinja
    alert_text: |
          The alert is generated between {{start_time}} to {{end_time}}
          Environemnt: Dev
          Rule-name: {{name}}
          Service: {{service}}
          routePattern: {{routePattern}}
          Backend_URL: {{http[0].url}}
          http.response code: {{http[0].response_code}}
          Backend_Total_time: {{http[0].total_time}}
          Error_code: {{http[0].error_code}}
          Error_Message: {{http[0].error}}
          TimeStamp: {{time}}
          Number_Hits: {{num_hits}}
          Number_Matches: {{num_matches}}

Could you please help me how can  i set this in alert_text.
Thanks
Ziaul

**Reply by jertel** [reply]
Please include the redacted debug logs showing that an alert was sent, and provide a redacted screenshot of the "blank" alert you received.

**Reply by ziaul-ict2020** [reply]
I sincerely apologize for that may I did not understand to you well.

My rule is working perfectly and getting the alert as per my expectation. But I want to include the following text into my alert_text when the alert will be triggered.

"At least 5 events occurred between 2023-02-27 04:53 UTC and 2023-02-27 05:53 UTC"


My question is that which variable will be i used  in my customize alert-text for the query period time.

My alert configuration as like below.

    alert:
    - "googlechat"
    googlechat_webhook_url: "https://chat.googleapis.com/v1/spaces/AAAAP_0p7aE/messages"
    googlechat_format: "card"
    googlechat_header_title: "Status_418_Without_Push_Notification"
    alert_text_type: alert_text_jinja
    alert_text: |
          At least 5 events occurred between {{starttime}} and {{endtime}}
          Environemnt: Production
          Rule-name: {{name}}
          Service: {{service}}
          routePattern: {{routePattern}}
          Backend_URL: {{http[0].url}}
          http.response code: {{http[0].response_code}}
          Backend_Total_time: {{http[0].total_time}}
          Error_code: {{http[0].error_code}}
          Error_Message: {{http[0].error}}
          TimeStamp: {{time}}
          Number_Hits: {{num_hits}}
          Number_Matches: {{num_matches}}

Here I have shared the log
Thanks

(log removed by jertel due to personal/commercial data)

**Reply by jertel** [reply]
The start/stop time variables do not exist in the `match` dict, which means they are unavailable to the Jinja template. In order to get that `At least ... events occurred between ... and ...` string to appear you would need to either:

1. Remove the `alert_text_type` and `alert_text` fields from your rule, which will cause ElastAlert 2 to generate its own default message.
or
2. Add a small Python enhancement module which populates the `starttime` and `endtime` key/values in the `match` dict. Search the docs and this discussion forum for `enhancement` to find more information and examples on this.
